compiler/src/iree/compiler/InputConversion/MHLO/ConvertCollectiveOps.cpp

```diff
@@ -577,13 +578,102 @@ struct ReduceScatterOpConversion
   }
 };
 
+struct CollectivePermuteOpConversion
+    : public OpConversionPattern<mhlo::CollectivePermuteOp> {
+  using OpConversionPattern<mhlo::CollectivePermuteOp>::OpConversionPattern;
+
+  LogicalResult matchAndRewrite(
+      mhlo::CollectivePermuteOp op, OpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const override {
+    auto loc = op.getLoc();
+
+    // Create a default channel.
+    auto channel = rewriter.create<IREE::Flow::ChannelDefaultOp>(
+        loc, /*group=*/StringAttr{});
+
+    auto inputType = op.getOperand().getType().cast<RankedTensorType>();
+
+    // Get the collective element type attribute.
+    IREE::Flow::CollectiveElementTypeAttr elementTypeAttr =
+        getCollectiveElementTypeAttr(op.getContext(), inputType);
+    if (!elementTypeAttr) {
+      return rewriter.notifyMatchFailure(op, "unsupported input type");
+    }
+
+    // Convert source target pairs into a constant table that can be indexed by
+    // rank to find which ids that rank should send to and recv from, or -1 for
+    // no send/recv.
+    DenseIntElementsAttr sourceTargetPairs = op.getSourceTargetPairs();
+    llvm::DenseMap<int64_t, int64_t> sendTable, recvTable;
+    int64_t maxRank = 0;
+    for (auto i = sourceTargetPairs.begin(); i != sourceTargetPairs.end();
+         ++i) {
+      int64_t source = (*i).getSExtValue();
+      int64_t target = (*++i).getSExtValue();
+      if (source > std::numeric_limits<int16_t>::max() ||
+          target > std::numeric_limits<int16_t>::max()) {
+        return rewriter.notifyMatchFailure(
+            op,
+            " source target pair id exceeds maximum value of 16-bit integer");
+      }
+      sendTable[source] = target;
+      recvTable[target] = source;
+      maxRank = std::max(std::max(maxRank, source), target);
+    }
+    SmallVector<uint32_t, 8> sendRecvTablePacked;
+    sendRecvTablePacked.reserve(maxRank + 1);
+    for (int64_t i = 0; i <= maxRank; ++i) {
+      int16_t sendId = sendTable.count(i) ? sendTable[i] : -1;
+      int16_t recvId = recvTable.count(i) ? recvTable[i] : -1;
+      int32_t packed;
+      std::memcpy(&packed, &sendId, 2);
+      std::memcpy(reinterpret_cast<char *>(&packed) + 2, &recvId, 2);
+      sendRecvTablePacked.push_back(packed);
+    }
+    // For rank ids > maxRank, we will clamp to this last entry which will
+    // return -1,-1. This is a workaround to avoid control flow since the number
+    // of ranks in the channel is not statically known.
+    sendRecvTablePacked.push_back(0xFFFFFFFF);
+
+    // Create constant to hold send/recv table..
```

Comment:
We may put this calculation into the runtime, then we don't need to create the boundary condition by just passing the pair attributes as is. In the runtime, we can simply iterate over the data to find the send and recv pair.

Reply:
I'm all for doing it that way, but I thought we didn't have a way to pass the `source_target_pairs` as-is to the HAL layer. Is it possible with the push buffer APIs and do you happen to have an example?

Reply:
@benvanik is looking into it to support the groups attribute in my PR for the group support. https://github.com/openxla/iree/pull/13089#issuecomment-1539265331

Reply:
I see, I can give it a try.
@benvanik any concerns about changing `param` in `hal.collective` and `stream.collective` to `Variadic<I32>`?

Reply:
I'd missed this was doing tensor stuff. We aren't going to want that - `tensor.extract` is _extremely_ expensive. `!util.buffer` is what we'd need to use instead for a host-side lookup table. You can use `util.buffer.constant` instead of `mhlo.constant` and `util.buffer.load` instead of `tensor.extract`.

We want to keep things like this in the compiler where possible as that allows for hoisting/CSE/etc - anything we leave to the runtime cannot be made any more efficient as it can't see the global scope. Shipping the entire table into the runtime and doing the lookup there also leaks and ossifies the implementation detail about how this actually works - by keeping it in the compiler we can support other inputs that may have different ways of doing this. We need to keep anything that is an HLO opinion at this layer of the stack.

Reply:
arith is ideal!